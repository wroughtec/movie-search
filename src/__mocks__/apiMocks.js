export const searchMock = {
  page: 1,
  total_results: 140,
  total_pages: 7,
  results: [
    {
      vote_count: 9703,
      id: 11,
      video: false,
      vote_average: 8.2,
      title: 'Star Wars',
      popularity: 44.515,
      poster_path: '/btTdmkgIvOi0FFip1sPuZI2oQG6.jpg',
      original_language: 'en',
      original_title: 'Star Wars',
      genre_ids: [],
      backdrop_path: '/4iJfYYoQzZcONB9hNzg0J0wWyPH.jpg',
      adult: false,
      overview:
        'Princess Leia is captured and held hostage by the evil Imperial forces in their effort to take over the galactic Empire. Venturesome Luke Skywalker and dashing captain Han Solo team together with the loveable robot duo R2-D2 and C-3PO to rescue the beautiful princess and restore peace and justice in the Empire.',
      release_date: '1977-05-25'
    }
  ]
};

export const configMock = {
  images: {
    base_url: 'http://image.tmdb.org/t/p/',
    secure_base_url: 'https://image.tmdb.org/t/p/',
    backdrop_sizes: ['w300', 'w780', 'w1280', 'original'],
    logo_sizes: ['w45', 'w92', 'w154', 'w185', 'w300', 'w500', 'original'],
    poster_sizes: ['w92', 'w154', 'w185', 'w342', 'w500', 'w780', 'original'],
    profile_sizes: ['w45', 'w185', 'h632', 'original'],
    still_sizes: ['w92', 'w185', 'w300', 'original']
  },
  change_keys: []
};

export const genresMock = [{ id: 1, name: 'new' }];

export const movieDetails = {
  adult: false,
  backdrop_path: '/4iJfYYoQzZcONB9hNzg0J0wWyPH.jpg',
  belongs_to_collection: {
    id: 10,
    name: 'Star Wars Collection',
    poster_path: '/iTQHKziZy9pAAY4hHEDCGPaOvFC.jpg',
    backdrop_path: '/d8duYyyC9J5T825Hg7grmaabfxQ.jpg'
  },
  budget: 11000000,
  genres: [{ id: 12, name: 'Adventure' }, { id: 28, name: 'Action' }, { id: 878, name: 'Science Fiction' }],
  homepage: 'http://www.starwars.com/films/star-wars-episode-iv-a-new-hope',
  id: 11,
  imdb_id: 'tt0076759',
  original_language: 'en',
  original_title: 'Star Wars',
  overview:
    'Princess Leia is captured and held hostage by the evil Imperial forces in their effort to take over the galactic Empire. Venturesome Luke Skywalker and dashing captain Han Solo team together with the loveable robot duo R2-D2 and C-3PO to rescue the beautiful princess and restore peace and justice in the Empire.',
  popularity: 47.377,
  poster_path: '/btTdmkgIvOi0FFip1sPuZI2oQG6.jpg',
  production_companies: [
    { id: 1, logo_path: '/o86DbpburjxrqAzEDhXZcyE8pDb.png', name: 'Lucasfilm', origin_country: 'US' },
    { id: 25, logo_path: '/qZCc1lty5FzX30aOCVRBLzaVmcp.png', name: '20th Century Fox', origin_country: 'US' }
  ],
  production_countries: [{ iso_3166_1: 'US', name: 'United States of America' }],
  release_date: '1977-05-25',
  revenue: 775398007,
  runtime: 121,
  spoken_languages: [{ iso_639_1: 'en', name: 'English' }],
  status: 'Released',
  tagline: 'A long time ago in a galaxy far, far away...',
  title: 'Star Wars',
  video: false,
  vote_average: 8.2,
  vote_count: 9718
};
