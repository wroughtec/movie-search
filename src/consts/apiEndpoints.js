// @flow

export const searchEndpoint = 'search/movie';
export const configEndpoint = 'configuration';
export const genresEndpoint = 'genre/movie/list';
export const popularEndpoint = 'movie/popular';
export const movieDetailsEndpoint = (id: string) => `movie/${id}`;
