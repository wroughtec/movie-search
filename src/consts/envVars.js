// @flow

export const BASE_URL: ?string = process.env.API_URL;
export const KEY: ?string = process.env.API_KEY;
