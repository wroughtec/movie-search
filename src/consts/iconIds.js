// @flow

export const spinnerIcon = 'spinner';
