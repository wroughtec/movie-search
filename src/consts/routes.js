export const home = '/';
export const search = '/search/';
export const movie = '/movie/';
