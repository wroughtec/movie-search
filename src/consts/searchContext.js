export const searchContext = {
  loading: true,
  imageBaseUrl: '',
  sm: '',
  md: '',
  lg: '',
  xl: '',
  searchTerms: '',
  searchResults: {},
  handleMovieSubmit() {},
  handleSearchChange() {}
};
