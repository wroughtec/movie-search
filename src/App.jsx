// @flow
import React, { Component } from 'react';
import { produce } from 'immer';
import RequestMovies from 'utils/requestMovies';
import { Provider } from 'components/SearchContext/SearchContext';
import { ResultsWrapper } from 'components/Results/ResultsWrapper';
import { SearchBox } from 'components/SearchBox/SearchBox';
import { Nav } from 'components/Nav/Nav';
import { Router, navigate } from '@reach/router';
import { home, search } from 'consts/routes';
import 'app/_app.scss';

type State = {
  ...SearchContextType
};

export class App extends Component<void, State> {
  constructor(props) {
    super(props);

    this.state = {
      loading: true,
      imageBaseUrl: '',
      sm: '',
      md: '',
      lg: '',
      xl: '',
      genres: [],
      searchTerms: '',
      searchResults: {},
      searchPageNo: 1,
      popularResults: {},
      popularPageNo: 1,
      handleSearchChange: this.handleSearchChange,
      handleMovieSubmit: this.handleMovieSubmit,
      updateSearchTerms: this.updateSearchTerms,
      loadMovies: this.loadMovies
    };
  }

  async componentDidMount() {
    Promise.all([this.addConfigToState(), this.addGenresToState(), this.addPopularToState()]).then(() =>
      this.loadingComplete()
    );
  }

  loadingComplete = () => {
    this.setState(
      produce(this.state, draft => {
        draft.loading = false;
      })
    );
  };

  addPopularToState = async () => {
    const popular = await RequestMovies.popularMovies();

    if (popular && popular.results) {
      this.setState(
        produce(this.state, draft => {
          draft.popularResults = popular;
        })
      );
    }
  };

  addConfigToState = async () => {
    const config = await RequestMovies.config();

    let sm, md, lg, xl, imageBaseUrl;

    if (config && config.images) {
      imageBaseUrl = config.images.secure_base_url;
      [sm, md, lg, xl] = config.images.still_sizes;

      this.setState(
        produce(this.state, draft => {
          draft.imageBaseUrl = imageBaseUrl;
          draft.sm = sm;
          draft.md = md;
          draft.lg = lg;
          draft.xl = xl;
        })
      );
    }
  };

  addGenresToState = async () => {
    const genres = await RequestMovies.getGenres();

    if (Array.isArray(genres.genres)) {
      this.setState(
        produce(this.state, draft => {
          draft.genres = genres.genres;
        })
      );
    }
  };

  handleSearchChange = (event: SyntheticEvent<HTMLButtonElement>) => {
    const searchTerms = event.currentTarget.value;

    this.updateSearchTerms(searchTerms);
  };

  updateSearchTerms = (searchTerms: string) => {
    this.setState(
      produce(this.state, draft => {
        draft.searchTerms = searchTerms;
      })
    );
  };

  handleMovieSubmit = (event: SyntheticEvent<HTMLButtonElement>) => {
    event.preventDefault();
    const { searchTerms } = this.state;
    navigate(`/search/${searchTerms}`);
    this.loadMovies();
  };

  loadMovies = () => {
    this.setState(
      produce(this.state, draft => {
        draft.loading = true;
      }),
      this.searchForMovies
    );
  };

  searchForMovies = async () => {
    const { searchTerms, searchPageNo } = this.state,
      searchResults = await RequestMovies.searchMovies(searchTerms, searchPageNo);

    this.setState(
      produce(this.state, draft => {
        draft.searchResults = searchResults;
        draft.loading = false;
      })
    );
  };

  render() {
    return (
      <Provider value={this.state}>
        <SearchBox />
        <Nav resetSearchTerm={this.updateSearchTerms} />
        <Router>
          <ResultsWrapper path={home} popular />
          <ResultsWrapper path={search} />
        </Router>
      </Provider>
    );
  }
}
