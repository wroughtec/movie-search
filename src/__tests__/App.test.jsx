import React from 'react';
import fetchMock from 'fetch-mock';
import { render } from 'react-dom';
import { shallow } from 'enzyme';
import { App } from 'app/App';
import { searchEndpoint, genresEndpoint, configEndpoint, popularEndpoint } from 'consts/apiEndpoints';
import { BASE_URL, KEY } from 'consts/envVars';
import { searchMock, configMock, genresMock } from 'mocks/apiMocks';

describe('App', () => {
  fetchMock.get(`${BASE_URL}/${genresEndpoint}?api_key=${KEY}`, JSON.stringify({ genres: genresMock }));
  fetchMock.get(`${BASE_URL}/${configEndpoint}?api_key=${KEY}`, JSON.stringify({ ...configMock }));
  fetchMock.get(`${BASE_URL}/${searchEndpoint}?api_key=${KEY}`, JSON.stringify({ ...searchMock }));
  fetchMock.get(`${BASE_URL}/${popularEndpoint}?api_key=${KEY}`, JSON.stringify({ ...searchMock }));
  fetchMock.get('*', JSON.stringify({ error: 'error' }));

  const event = { preventDefault: () => {} },
    currentTarget = {
      currentTarget: {
        value: 'star wars'
      }
    },
    sleep = ms => new Promise(resolve => setTimeout(resolve, ms)),
    DELAY_MS = 2000;

  beforeEach(() => {
    jest.spyOn(event, 'preventDefault');
  });

  it('renders without crashing', () => {
    const div = document.createElement('div');

    render(<App />, div);
  });

  it('should render correctly', () => {
    const component = shallow(<App />);

    expect(component).toMatchSnapshot();
  });

  test(`loads data into state in componentDidMount`, async () => {
    const wrapper = shallow(<App />);

    await wrapper.instance().componentDidMount();
    // Add fake delay
    await sleep(DELAY_MS);
    expect(wrapper.state('imageBaseUrl')).toEqual('https://image.tmdb.org/t/p/');
    expect(wrapper.state('genres')).toEqual(genresMock);
    expect(wrapper.state('popularResults')).toEqual(searchMock);

    wrapper.update();
    expect(wrapper.find('SearchBox')).toHaveLength(1);
    expect(wrapper.find('ResultsWrapper')).toHaveLength(2);
  });

  test('handleSubmit', async () => {
    const wrapper = shallow(<App />);

    await wrapper.instance().handleMovieSubmit(event);
    await sleep(DELAY_MS);
    expect(wrapper.state('searchResults')).toEqual(searchMock);
  });

  test('handleOnChange', async () => {
    const wrapper = shallow(<App />);

    await wrapper.instance().handleSearchChange(currentTarget);
    await sleep(DELAY_MS);
    expect(wrapper.state('searchTerms')).toEqual('star wars');
  });

  test('handleSubmit Error', async () => {
    const wrapper = shallow(<App />);
    expect.assertions(1);
    try {
      await wrapper.instance().handleMovieSubmit();
    } catch (e) {
      expect(e).toEqual(e);
    }
  });

  test('componentDidMount endpoints no data', async () => {
    fetchMock.config.overwriteRoutes = true;
    fetchMock.get(`${BASE_URL}/${configEndpoint}?api_key=${KEY}`, JSON.stringify({ error: 'error' }));
    fetchMock.get(`${BASE_URL}/${genresEndpoint}?api_key=${KEY}`, JSON.stringify({ error: 'error' }));
    fetchMock.get(`${BASE_URL}/${popularEndpoint}?api_key=${KEY}`, JSON.stringify({ error: 'error' }));
    const wrapper = shallow(<App />);
    await wrapper.instance().componentDidMount();
    // Add fake delay
    await sleep(DELAY_MS);

    expect(wrapper.state('imageBaseUrl')).toEqual('');
    expect(wrapper.state('genres')).toEqual([]);
    expect(wrapper.state('popularResults')).toEqual({});
  });
});
