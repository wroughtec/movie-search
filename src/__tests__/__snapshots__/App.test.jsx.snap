// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <SearchBox />,
        <Nav
          resetSearchTerm={[Function]}
        />,
        <Router>
          <ResultsWrapper
            path="/"
            popular={true}
          />
          <ResultsWrapper
            path="/search/:searchTerm"
          />
          <MovieWrapper
            path="/movie/:movieId"
          />
        </Router>,
      ],
      "value": Object {
        "genres": Array [],
        "handleMovieSubmit": [Function],
        "handleSearchChange": [Function],
        "imageBaseUrl": "",
        "lg": "",
        "loadMovies": [Function],
        "loading": true,
        "md": "",
        "popularPageNo": 1,
        "popularResults": Object {},
        "searchPageNo": 1,
        "searchResults": Object {},
        "searchTerms": "",
        "sm": "",
        "updateSearchTerms": [Function],
        "xl": "",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "resetSearchTerm": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <ResultsWrapper
              path="/"
              popular={true}
            />,
            <ResultsWrapper
              path="/search/:searchTerm"
            />,
            <MovieWrapper
              path="/movie/:movieId"
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "path": "/",
              "popular": true,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "path": "/search/:searchTerm",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "path": "/movie/:movieId",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": Object {
      "$$typeof": Symbol(react.provider),
      "_context": Object {
        "$$typeof": Symbol(react.context),
        "Consumer": [Circular],
        "Provider": [Circular],
        "_calculateChangedBits": null,
        "_currentRenderer": Object {},
        "_currentRenderer2": null,
        "_currentValue": Object {
          "genres": Array [],
          "handleMovieSubmit": [Function],
          "handleSearchChange": [Function],
          "imageBaseUrl": "",
          "lg": "",
          "loading": true,
          "md": "",
          "popularPageNo": 1,
          "popularResults": Object {},
          "searchPageNo": 1,
          "searchResults": Object {},
          "searchTerms": "",
          "sm": "",
          "updateSearchTerms": [Function],
          "xl": "",
        },
        "_currentValue2": Object {
          "genres": Array [],
          "handleMovieSubmit": [Function],
          "handleSearchChange": [Function],
          "imageBaseUrl": "",
          "lg": "",
          "loading": true,
          "md": "",
          "popularPageNo": 1,
          "popularResults": Object {},
          "searchPageNo": 1,
          "searchResults": Object {},
          "searchTerms": "",
          "sm": "",
          "updateSearchTerms": [Function],
          "xl": "",
        },
        "unstable_read": [Function],
      },
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <SearchBox />,
          <Nav
            resetSearchTerm={[Function]}
          />,
          <Router>
            <ResultsWrapper
              path="/"
              popular={true}
            />
            <ResultsWrapper
              path="/search/:searchTerm"
            />
            <MovieWrapper
              path="/movie/:movieId"
            />
          </Router>,
        ],
        "value": Object {
          "genres": Array [],
          "handleMovieSubmit": [Function],
          "handleSearchChange": [Function],
          "imageBaseUrl": "",
          "lg": "",
          "loadMovies": [Function],
          "loading": true,
          "md": "",
          "popularPageNo": 1,
          "popularResults": Object {},
          "searchPageNo": 1,
          "searchResults": Object {},
          "searchTerms": "",
          "sm": "",
          "updateSearchTerms": [Function],
          "xl": "",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "resetSearchTerm": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <ResultsWrapper
                path="/"
                popular={true}
              />,
              <ResultsWrapper
                path="/search/:searchTerm"
              />,
              <MovieWrapper
                path="/movie/:movieId"
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "path": "/",
                "popular": true,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "path": "/search/:searchTerm",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "path": "/movie/:movieId",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": Object {
        "$$typeof": Symbol(react.provider),
        "_context": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": [Circular],
          "_calculateChangedBits": null,
          "_currentRenderer": Object {},
          "_currentRenderer2": null,
          "_currentValue": Object {
            "genres": Array [],
            "handleMovieSubmit": [Function],
            "handleSearchChange": [Function],
            "imageBaseUrl": "",
            "lg": "",
            "loading": true,
            "md": "",
            "popularPageNo": 1,
            "popularResults": Object {},
            "searchPageNo": 1,
            "searchResults": Object {},
            "searchTerms": "",
            "sm": "",
            "updateSearchTerms": [Function],
            "xl": "",
          },
          "_currentValue2": Object {
            "genres": Array [],
            "handleMovieSubmit": [Function],
            "handleSearchChange": [Function],
            "imageBaseUrl": "",
            "lg": "",
            "loading": true,
            "md": "",
            "popularPageNo": 1,
            "popularResults": Object {},
            "searchPageNo": 1,
            "searchResults": Object {},
            "searchTerms": "",
            "sm": "",
            "updateSearchTerms": [Function],
            "xl": "",
          },
          "unstable_read": [Function],
        },
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
